### DISCUSSION TITLE
Accessing `config` in `processFeature` or `preprocessOsmRelation`

### DISCUSSION THREAD

**Discussion by zstadler** [onthegomap/planetiler#957] [Q&A] [answered]
Thanks @msbarry for the help [on Slack](https://osmus.slack.com/archives/C031V9E9RMG/p1718982618982259) regarding the creation of a point layer for OSM route relations.

In summary, the approach suggested on Slack is to save all the required information on the relation during the `preprocessOsmRelation` stage, including the smallest id of its way members. Once such a member way is detected during the `processFeature` stage, a point tile-feature is added (at the start location of the way), with the attributes saved from the relation's tags.

The missing piece in this puzzle is how to use the relation's OSM id as the basis for the feature id of that point.

Any `sourceFeature`, such as the relation, has a `vectorTileFeatureId(int multiplier)` method that computes the id for the tile-feature, where the multiplier is taken from `config`, as done in [`FeatureCollector`](https://github.com/onthegomap/planetiler/blob/8ffe290c229e05ed34964bc10c861954422953fa/planetiler-core/src/main/java/com/onthegomap/planetiler/FeatureCollector.java#L69-L70) 

```
    long vectorTileId = config.featureSourceIdMultiplier() < 4 ? source.id() :
      source.vectorTileFeatureId(config.featureSourceIdMultiplier());
```

I was unable to fine a way to access the config from `preprocessOsmRelation` or `processFeature`. As a workaround, I've re-coded the formula using the default `config.featureSourceIdMultiplier() == 10`

```
features.geometry(...)
   .setAttr(...)
   ...
   .setId(relation.id() * 10 + 3);
```

How can I implement this using `featureSourceIdMultiplier()` and `vectorTileFeatureId()`?

**Comment by msbarry** [comment]
Can you pass config into the constructor for the layer and store it in a field to use in processFeature or other profile methods?

**Reply by zstadler** [reply]
My current setup is similar to the bike routes example.

I've passed `config` into the profile in `run()`
```
    Planetiler planetiler = Planetiler.create(args);
    config = planetiler.config();
    planetiler.setProfile(new TrailPois())
      // override this default with osm_path="path/to/data.osm.pbf"
      .addOsmSource("osm", Path.of("data", "sources", area + ".osm.pbf"), "geofabrik:" + area)
      // override this default with mbtiles="path/to/output.mbtiles"
      .overwriteOutput(Path.of("data", "trail_pois.pmtiles"))
      .run();
  }
```

The next problem is that `vectorTileFeatureId` is a method of `SourceFeature` while `preprocessOsmRelation` receives a `OsmElement.Relation` object that does not have that method.

**Reply by msbarry** [reply, marked as answer]
Looks like the API doesn't expose this too easily, would this let you do what you need? https://github.com/onthegomap/planetiler/pull/960/files - seems like you could call `feature.setId(relation.vectorTileFeatureId(multiplier))` ?

**Reply by zstadler** [reply]
Thanks! It looks like that would solve the issue.
I'll try to build the `relation-id` branch locally and test my code with it. It may take a few days until I'll be able to get to it.

**Reply by zstadler** [reply]
Thank you @msbarry!
It works like charm!

**Reply by msbarry** [reply]
OK great! Want me to just merge that PR then?

**Reply by zstadler** [reply]
Definitely yes!
Please merge this PR so I can commit my code that uses it.

**Reply by msbarry** [reply]
sounds good, just merged it
